### PR TITLE
ITE-7 Proposal Draft: Add support for PKI signing & verification

### PR DESCRIPTION
**Pull request by mikhailswift**
This is a rough draft of an ITE detailing the work we've done at https://github.com/boxboat/in-toto-golang

We aim to support signing and verification using short-lived keys & certificates from existing PKI.
Our specific use case involved keys coming from a SPIFFE/SPIRE trust domain but this should
be usable by any PKI setup, short lived or not.  

This will require changes if ITE6 is adopted and a discussion is being held in the DSSE repository
about how to merge these ideas.

### CODE REVIEW

**Comment by trishankatdatadog** [comment]
Looks interesting! Happy to review. When is a good deadline?

File: ITE/7/README.adoc [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+     "public": ""
+    }
+}
+```
+
+During the verification process a functionary will be tested against the
+existing set of known key's as defined in the layout or whether a chain of trust
+can be built to the defined roots using the defined intermediates.
+
+Additonal intermediate could be supplied at time of verification though security
```
Comment (MarkLodato) [inline ITE/7/README.adoc:87]: typo: Additional

File: ITE/7/README.adoc [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+    "keyval": {
+     "certificate": "-----BEGIN CERTIFICATE-----\nMIIBkjCCATegAwIBAgIBADAKBggqhkjOPQQDAjAdMQswCQYDVQQGEwJVUzEOMAwG\nA1UEChMFU1BJUkUwHhcNMjEwMzAzMTk0MjI0WhcNMjEwNDAyMTk0MjM0WjAdMQsw\nCQYDVQQGEwJVUzEOMAwGA1UEChMFU1BJUkUwWTATBgcqhkjOPQIBBggqhkjOPQMB\nBwNCAARbJaNMniz2ejaGwLAS5Kfl3modn0ceD6LXw+QltwIJKIqGO3C8Lh2KGmZ+\nBycxOHpDcHky8NMdM+0dIVawlIlVo2gwZjAOBgNVHQ8BAf8EBAMCAYYwDwYDVR0T\nAQH/BAUwAwEB/zAdBgNVHQ4EFgQU0dLhyMLPbujKf9nW7j/7qUheP7IwJAYDVR0R\nBB0wG4YZc3BpZmZlOi8vc3BpcmUuYm94Ym9hdC5pbzAKBggqhkjOPQQDAgNJADBG\nAiEA4RYLyrSxwUbv3h1X8kpfyLQmOniCbbMZqvIS49GcWtMCIQD309bBx89ITsYx\nxskO9LGz7NM1QYeiETY3LgZ6joIdgg==\n-----END CERTIFICATE-----\n",
+     "private": "",
+     "public": ""
+    }
+}
+```
+
+During the verification process a functionary will be tested against the
+existing set of known key's as defined in the layout or whether a chain of trust
```
Comment (MarkLodato) [inline ITE/7/README.adoc:84]: typo: keys

**Review by MarkLodato — COMMENTED** [review]
Add link from the top-level README.md.

(No comment on content - I'm not enough of an in-toto expert to offer an opinion.)

File: ITE/7/README.adoc [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+can be built to the defined roots using the defined intermediates.
+
+Additonal intermediate could be supplied at time of verification though security
+concerns discussed below may be undesired.
```
Comment (adityasaky) [inline ITE/7/README.adoc:88]:
```suggestion
concerns discussed below may deem this undesirable.
```

File: ITE/7/README.adoc [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+[[prototype-implementation]]
+== Prototype Implementation
+
+A current proof-of-concept implementation of this ITE exists at Boxboat's fork
+of the in-toto-golang project: https://github.com/boxboat/in-toto
+
+[[references]]
+== References
+
+=== example references
```
Comment (adityasaky) [inline ITE/7/README.adoc:257]:
```suggestion
```

**Review by adityasaky — COMMENTED** [review]
Hi @mikhailswift, thanks for submitting this work as an ITE! I've left some mostly administrative comments below. Could you also add the ITE to the readme?

**Comment by mikhailswift** [comment]
Went ahead and made the suggested edits :)  thanks for the review!

**Comment by trishankatdatadog** [comment]
> We need something like this for sigstore/fulcio trust roots.

Aren't we using [SigStore Trust Delegations](https://docs.google.com/document/d/1lNW2lNu9QmDhCgXJaNTYxBUzCjuWS2wj8fCHSadbzxQ/edit?usp=sharing) to bootstrap trust? @asraa

**Comment by lukehinds** [comment]
> > We need something like this for sigstore/fulcio trust roots.
> 
> Aren't we using [SigStore Trust Delegations](https://docs.google.com/document/d/1lNW2lNu9QmDhCgXJaNTYxBUzCjuWS2wj8fCHSadbzxQ/edit?usp=sharing) to bootstrap trust? @asraa

I worded my thing badly. I meant leveraging the OIDC email addr as a trust root (well more of a policy root, alongside the x509 trust to the fulcio ca root).

File: ITE/7/README.adoc [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+to support verifying functionaries with a chain of trust to a known root
+certificate authority.
+
+Modifications to the layout include the addition of a `rootcas` and
+`intermediatecas` field.  These fields allow the creator of the layout to
+specify known trusted certificate authorities that a functionary's signing key
+must belong to.
+
+Additionally a `cert_constraints` field is added to steps within a layout.  This
+field allows the layout creator to limit the keys a functionary by some
```
Comment (asraa) [inline ITE/7/README.adoc:58]: having some trouble parsing:
` allows the layout creator to limit the keys a functionary `

> allows the layout creator to limit the functionary keys by some attributes?

File: ITE/7/README.adoc [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+attributes of the key's certificate, ensuring the key has a specific common
+name or URI, for instance.
+
+[[authorities]]
+=== Certificate Authorities
+
+The `intermediatecas` and `rootcas` fields will have the same shape as the
+existing `keys` field: an object that with string indexes that are equal to the
+key's ID.  The keyval's public key will be expected to be a valid X509
+certificate.
```
Comment (asraa) [inline ITE/7/README.adoc:68]: PEM encoded? or is that up to the implementer and outside of the scope of this Proposal

File: ITE/7/README.adoc [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+     "public": ""
+    }
+}
+```
+
+During the verification process a functionary will be tested against the
+existing set of known keys as defined in the layout or whether a chain of trust
+can be built to the defined roots using the defined intermediates.
+
+Additional intermediate could be supplied at time of verification though security
```
Comment (asraa) [inline ITE/7/README.adoc:87]: intermediates*

**Comment by SantiagoTorres** [comment]
Discussed in February Community Meeting. Merging as draft!

**Pull request merged**